api/prometheus.go

```diff
@@ -0,0 +1,453 @@
+
+	q := querier{
+		*s,
+		uint32(start.Unix()),
+		uint32(end.Unix()),
+		ctx.OrgId,
+		true,
+		newCtx,
+	}
+	if err != nil {
+		response.Write(ctx, response.NewError(http.StatusInternalServerError, fmt.Sprintf("query failed: %v", err)))
+		return
+	}
+	var matcherSets [][]*labels.Matcher
+	for _, s := range request.Match {
+		matchers, err := promql.ParseMetricSelector(s)
+		if err != nil {
+			response.Write(ctx, response.NewJson(http.StatusInternalServerError, prometheusQueryResult{
+				Status:    statusError,
+				Error:     fmt.Errorf("query failed: %v", err),
+				ErrorType: errorBadData,
+			}, ""))
+			return
+		}
+		matcherSets = append(matcherSets, matchers)
+	}
+
+	var sets []storage.SeriesSet
+	for _, mset := range matcherSets {
+		s, err := q.Select(mset...)
+		if err != nil {
+			response.Write(ctx, response.NewJson(http.StatusInternalServerError, prometheusQueryResult{
+				Status:    statusError,
+				Error:     fmt.Errorf("query failed: %v", err),
+				ErrorType: errorExec,
+			}, ""))
+			return
+		}
+		sets = append(sets, s)
+	}
+
+	set := storage.NewMergeSeriesSet(sets)
+	metrics := []labels.Labels{}
+	for set.Next() {
+		metrics = append(metrics, set.At().Labels())
+	}
+	if set.Err() != nil {
+		response.Write(ctx, response.NewJson(http.StatusInternalServerError, prometheusQueryResult{
+			Status:    statusError,
+			Error:     fmt.Errorf("query failed: %v", err),
+			ErrorType: errorExec,
+		}, ""))
+		return
+	}
+
+	response.Write(ctx, response.NewJson(200,
+		prometheusQueryResult{
+			Data:   metrics,
+			Status: statusSuccess,
+		},
+		"",
+	))
+
+	return
+}
+
+func (s *Server) Querier(ctx context.Context, mint, maxt int64) (storage.Querier, error) {
+	return &querier{
+		*s,
+		uint32(mint / 1000), //Convert from NS to S
+		uint32(maxt / 1000), //TODO abstract this out into a separate function that is more accurate
+		ctx.Value(orgID("org-id")).(int),
+		false,
+		ctx,
+	}, nil
+}
+
+func parseTime(s string) (time.Time, error) {
+	if t, err := strconv.ParseFloat(s, 64); err == nil {
+		s, ns := math.Modf(t)
+		return time.Unix(int64(s), int64(ns*float64(time.Second))), nil
+	}
+	if t, err := time.Parse(time.RFC3339Nano, s); err == nil {
+		return t, nil
+	}
+	return time.Time{}, fmt.Errorf("cannot parse %q to a valid timestamp", s)
+}
+
+func parseDuration(s string) (time.Duration, error) {
+	if d, err := strconv.ParseFloat(s, 64); err == nil {
+		ts := d * float64(time.Second)
+		if ts > float64(math.MaxInt64) || ts < float64(math.MinInt64) {
+			return 0, fmt.Errorf("cannot parse %q to a valid duration. It overflows int64", s)
+		}
+		return time.Duration(ts), nil
+	}
+	if d, err := model.ParseDuration(s); err == nil {
+		return time.Duration(d), nil
+	}
+	return 0, fmt.Errorf("cannot parse %q to a valid duration", s)
+}
+
+// Select returns a set of series that matches the given label matchers.
+func (q *querier) Select(matchers ...*labels.Matcher) (storage.SeriesSet, error) {
```

Comment:
these querier functions should be declared under the querier type and under the NewQuerier constructor. in fact, probably cleaner to create a prometheus_querier.go file for it